src/probnum/kernels/_kernel.py

```diff
@@ -140,135 +364,99 @@ def _check_and_reshape_inputs(
+            raise ValueError(err_msg.format(argname="x0", shape=x0.shape))
+
+        broadcast_input_shape = x0.shape
+
+        if x1 is not None:
+            if x1.shape[-1] not in (self.input_dim, 1):
+                # This will never be called if the original input was a scalar
+                raise ValueError(err_msg.format(argname="x1", shape=x1.shape))
+
+            try:
+                # Ironically, `np.broadcast_arrays` seems to be more efficient than
+                # `np.broadcast_shapes`
+                broadcast_input_shape = np.broadcast_arrays(x0, x1)[0].shape
+            except ValueError as v:
                 raise ValueError(
-                    f"Argument shape x0.shape={x0.shape} does not match "
-                    "kernel input dimension."
-                )
+                    f"The input arrays `x0` and `x1` with shapes {x0.shape} and "
+                    f"{x1.shape} can not be broadcast to a common shape."
+                ) from v
 
-            # Determine correct shape for the kernel matrix as the output of __call__
-            kernshape = self._get_shape_kernelmatrix(
-                x0_shape=x0.shape, x1_shape=x0.shape
-            )
+        if broadcast_input_shape[-1] == 1:
+            broadcast_input_shape = broadcast_input_shape[:-1] + (self.input_dim,)
 
-            return np.atleast_2d(x0), None, kernshape
-        else:
-            x1 = np.asarray(x1)
-            err_msg = (
-                f"Argument shapes x0.shape={x0.shape} and x1.shape="
-                f"{x1.shape} do not match kernel input dimension "
-                f"{self.input_dim}. Try passing either two vectors or two "
-                "matrices with the second dimension equal to the kernel input "
-                "dimension."
-            )
+        return broadcast_input_shape
 
-            # Promote unequal shapes
-            if x0.ndim < 2 and x1.ndim == 2:
-                x0 = np.atleast_2d(x0)
-            if x1.ndim < 2 and x0.ndim == 2:
-                x1 = np.atleast_2d(x1)
-            if x0.ndim != x1.ndim:  # Shape mismatch
-                raise ValueError(err_msg)
-
-            # Check shapes
-            if (
-                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
-                or (
-                    x0.ndim == 1  # Vector input
-                    and not (x0.shape[0] == x1.shape[0] == self.input_dim)
-                )
-                or (
-                    x0.ndim == 2  # Matrix input
-                    and not (x0.shape[1] == x1.shape[1] == self.input_dim)
-                )
-            ):
-                raise ValueError(err_msg)
-
-            # Determine correct shape for the kernel matrix as the output of __call__
-            kernshape = self._get_shape_kernelmatrix(
-                x0_shape=x0.shape, x1_shape=x1.shape
-            )
+    def _euclidean_inner_products(
+        self, x0: np.ndarray, x1: Optional[np.ndarray]
```

Comment:
```suggestion
        self, x0: np.ndarray, x1: Optional[np.ndarray] = None
```
Again. Is there a reason this is not `ArrayLike`?

Reply:
See above